test/envtest/kongpluginbinding_managed_test.go

```diff
@@ -0,0 +1,240 @@
+				},
+			},
+			nil,
+		)
+
+	reconcilers := []Reconciler{
+		konnect.NewKongPluginReconciler(false, mgr.GetClient()),
+		konnect.NewKonnectEntityReconciler(factory, false, mgr.GetClient(),
+			konnect.WithKonnectEntitySyncPeriod[configurationv1alpha1.KongPluginBinding](konnectSyncTime),
+		),
+		konnect.NewKonnectEntityReconciler(factory, false, mgr.GetClient(),
+			konnect.WithKonnectEntitySyncPeriod[configurationv1alpha1.KongService](konnectSyncTime),
+		),
+	}
+
+	StartReconcilers(ctx, t, mgr, logs, reconcilers...)
+
+	rateLimitingkongPlugin := &configurationv1.KongPlugin{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: "rate-limiting-kp-",
+		},
+		PluginName: "rate-limiting",
+		Config: apiextensionsv1.JSON{
+			Raw: []byte(`{"minute": 5, "policy": "local"}`),
+		},
+	}
+	require.NoError(t, clientNamespaced.Create(ctx, rateLimitingkongPlugin))
+	t.Logf("deployed %s KongPlugin (%s) resource", client.ObjectKeyFromObject(rateLimitingkongPlugin), rateLimitingkongPlugin.PluginName)
+
+	kongserviceName := "kongservice-1"
+	kongService := &configurationv1alpha1.KongService{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: kongserviceName,
+			Annotations: map[string]string{
+				consts.PluginsAnnotationKey: rateLimitingkongPlugin.Name,
+			},
+		},
+		Spec: configurationv1alpha1.KongServiceSpec{
+			ControlPlaneRef: &configurationv1alpha1.ControlPlaneRef{
+				Type: configurationv1alpha1.ControlPlaneRefKonnectNamespacedRef,
+				KonnectNamespacedRef: &configurationv1alpha1.KonnectNamespacedRef{
+					Name: cp.Name,
+				},
+			},
+			KongServiceAPISpec: configurationv1alpha1.KongServiceAPISpec{
+				Name: lo.ToPtr(kongserviceName),
+				URL:  lo.ToPtr("http://example.com"),
+			},
+		},
+	}
+	require.NoError(t, clientNamespaced.Create(ctx, kongService))
+	t.Logf("deployed %s KongService resource", client.ObjectKeyFromObject(kongService))
+
+	pluginID := uuid.NewString()
+	factory.SDK.PluginSDK.EXPECT().
+		CreatePlugin(mock.Anything, mock.Anything, mock.Anything).
+		Return(
+			&sdkkonnectops.CreatePluginResponse{
+				Plugin: &sdkkonnectcomp.Plugin{
+					ID: lo.ToPtr(pluginID),
+				},
+			},
+			nil,
+		)
+	factory.SDK.PluginSDK.EXPECT().
+		UpsertPlugin(mock.Anything, mock.Anything, mock.Anything).Maybe().
+		Return(
+			&sdkkonnectops.UpsertPluginResponse{
+				Plugin: &sdkkonnectcomp.Plugin{
+					ID: lo.ToPtr(pluginID),
+				},
+			},
+			nil,
+		)
+
+	t.Logf("waiting for KongPluginBinding to be created")
+	w, err := clientWithWatch.Watch(ctx, &configurationv1alpha1.KongPluginBindingList{}, client.InNamespace(ns.Name))
+	require.NoError(t, err)
+	t.Cleanup(func() { w.Stop() })
+	kongPluginBinding := watchFor(t, ctx, w, watch.Added,
+		func(kpb *configurationv1alpha1.KongPluginBinding) bool {
+			return kpb.Spec.Targets.ServiceReference != nil &&
+				kpb.Spec.Targets.ServiceReference.Name == kongService.Name &&
+				kpb.Spec.PluginReference.Name == rateLimitingkongPlugin.Name
+		},
+		"KongPluginBinding wasn't created",
+	)
```

Comment:
Question out of curiosity: was there any specific reason not to rely on the usual way we handle eventual conditions (`require.Eventually`) but using `Watch`?

Reply:
I wanted to test this out as it feels more natural (might need some refactoring love though): we get notified as soon as something happens, not based on a time period.

Reply:
I understand it can be better in terms of the time spent on waiting as we receive an event as soon as it happens instead of using a fixed interval. I'm just wondering if we'd like to switch to this approach to not have a mix of both in the repo.

Reply:
I'm happy to hear feedback on this approach.